README.md
# ZeverCloud Solar Power for Homey

When you connect your ZeverSolar device to the ZeverCloud you can use this homey app for retrieving the 5 minute values of your solar panels and use this in your Homey home automation flows. Like charging your car or start your washing machine when there is enough solar power.

## Device

Every ZeverSolar plant is a device which you can monitor in homey.

### Requesting the app and secret key from zeversolar

Mail to service.eu@zeversolar.net , the app & key fields are hidden fields on your zevercloud account

### Retrieving the api key of a plant

When you go this page https://www.zevercloud.com/station/update and add step 5 you can get the api key of your plant. Use this api key as input for this app.

## Flows

### triggers

- PowerAbove100W
- PowerAbove500W
- PowerAbove1000W
- PowerIs0W

### conditions

- is_generating
- generating_output
  - minimal, under 100W
  - ok, between 100 & 500W"
  - better, between 500 & 1000W
  - nice, between 1000 & 1500W
  - super, between 1500 & 2000W
  - chill, between 2000 & 3000W
  - awesome, between 3000 & 4000W

CHANGES.md
# zevercloud app

## 0.0.20

- use new api endpoint plus request your app and secret key

## 0.0.7

- Support generating power from homey 3.0
- renaming measure_power-current to measure_power probably requires a reinstall of the device

## 0.0.6

- fix double device and cron tasks on my side. wait for https://github.com/athombv/homey-apps-sdk-issues/issues/50

## 0.0.5

- adjust e-total,e-month when unit = MWh

## 0.0.4

- adjust current power when it is above 1000W

## 0.0.3

- add insights

## 0.0.2

- intial version

drivers/index.js

(function () {

    // main settings
    const crypto = require('crypto');
    const https = require('http')

    var zevercloud = exports;

    // active functions()  -------------------------------------  active functions()  --------------------------------------------

    zevercloud.getTodayData = function getTodayData(settings) {
        console.log("node_modules settings " +  JSON.stringify(settings));
        
        var appSecret = settings["secret"]; 
        const appKey  = settings["appkey"];
        const apiKey  = settings["apikey"];
        
        const op = 'GET';
        const accept = 'application/json';
        const time = Date.now();
        
        const payload = op.concat('\n', accept,'\n\n\n',
                              '\nX-Ca-Key:', appKey,
                              '\nX-Ca-Timestamp:', time,
                              '\n/getPlantOverview?key=', apiKey );
        
        console.log(payload);
        
        const signedHeaders = "X-Ca-Signature-Headers: X-Ca-Key,X-Ca-Timestamp"
        
        let signature = crypto.createHmac('sha256', appSecret).update(payload).digest("base64")
        
        console.log(signature);
        
        const options = {
            protocol: 'http:',
            method: 'GET',
            headers: {
                'X-Ca-Timestamp': time,
                'Accept': accept,
                'X-Ca-Key': appKey,
                'X-Ca-Signature': signature,
                'X-Ca-Signature-Headers': 'X-Ca-Key,X-Ca-Timestamp'
            }
        }
        options.hostname = 'api.general.zevercloud.cn';
        options.path = '/getPlantOverview?key='+ apiKey;

        // let url = '/api/v1/getPlantOverview?key=' + settings["apikey"];

        return new Promise((resolve, reject) => {

            const reqX = https.request(options, resX => {
                console.log('-------------------');
                console.log(`statusCode: ${resX.statusCode}`);
                console.log(resX.headers);
            
                let body = "";
                let response = "";
                resX.on("data", data => {
                  body += data;
                });
                resX.on("end", () => {
                    // handle this
                    try {
                      response = JSON.parse(body.toString());
                    }
                    catch(error) {
                      response = "ERROR";
                    }
                    console.log(response);
                    resolve(response);
                });
            })  
            
            reqX.on('error', error => {
                console.error(error)
                reject(error);
            })
            reqX.end()

        });
    }
})();


drivers/plant/driver.js
'use strict';
'use math';

const Homey = require('homey');

class plantDriver extends Homey.Driver {
	
	onInit() {
		this.log('Plant driver has been inited');
	}

	onPair(socket) {
		var plantId = "dummy";
		// dummy
        let devices = [
            {
                "name": "zeversolar_plant_" + plantId,
                "data": { 
                    "id": plantId, 
                    "apikey": "...", 
                    "appkey": "...",
                    "secret": "..."                    
                },
                "settings": {
                    "id": plantId,
                    "apikey": "APIKey",
                    "appkey": "APPKey",
                    "secret": "Secret"   
                }
            }
        ]

        // this is called when the user presses save settings button in pair.html
        socket.on('get_devices', (device_data, callback) => {
            devices = device_data;
            callback(null, devices);
        });

        // this happens when user clicks away the pairing windows
        socket.on('disconnect', () => {
            this.log("ZeverCloud - Pairing is finished (done or aborted) ");
        })

    } // end onPair
}

module.exports = plantDriver;

drivers/plant/device.js
'use strict';

const zevercloud = require('../index.js');
const Homey = require('homey');

Date.prototype.timeNow = function(){ 
    return ((this.getHours() < 10)?"0":"") + ((this.getHours()>12)?(this.getHours()-12):this.getHours()) +":"+ ((this.getMinutes() < 10)?"0":"") + this.getMinutes() + " " + ((this.getHours()>12)?('PM'):'AM');
};

class plant extends Homey.Device {

	async onInit() {
		this.log('plant has been inited');
        let settings = this.getSettings();
        let name = 'zeversolar_plant_' + this.getData().id;
        let cronName = name.toLowerCase();
        this.log("device settings " +  JSON.stringify(settings));

        Homey.ManagerCron.getTask(cronName)
            .then(task => {
                this.log("The task exists: " + cronName);
                this.log('Unregistering cron:', cronName);
                Homey.ManagerCron.unregisterTask(cronName, function (err, success) {});
                Homey.ManagerCron.registerTask(cronName, "*/5 * * * *", settings)
                .then(task => {
                    task.on('run', settings => this.pollSeverCloud(settings));
                })
                .catch(err => {
                    this.log('problem with registering cronjob: ${err.message}');
                });            
            })
            .catch(err => {
                if (err.code == 404) {
                    this.log("The task has not been registered yet, registering task: " + cronName);
                    Homey.ManagerCron.registerTask(cronName, "*/5 * * * *", settings)
                        .then(task => {
                            task.on('run', settings => this.pollSeverCloud(settings));
                        })
                        .catch(err => {
                            this.log('problem with registering cronjob: ${err.message}');
                        });
                } else {
                    this.log('other cron error: ${err.message}');
                }
            });

        this._flowTriggerPowerAbove100W = new Homey.FlowCardTrigger('PowerAbove100W').register();
        this._flowTriggerPowerAbove500W = new Homey.FlowCardTrigger('PowerAbove500W').register();
        this._flowTriggerPowerAbove1000W = new Homey.FlowCardTrigger('PowerAbove1000W').register();
        this._flowTriggerPowerIs0W = new Homey.FlowCardTrigger('PowerIs0W').register();
        this._conditionIsGenerating = new Homey.FlowCardCondition('is_generating').register().registerRunListener((args, state) => {
                var result = (this.getCapabilityValue('measure_power') > 0 ) 
                return Promise.resolve(result);
        });        
        this._conditionGeneratingOutput = new Homey.FlowCardCondition('generating_output').register().registerRunListener((args, state) => {
            var result = (this.conditionIsGeneratingToString(this.getCapabilityValue('measure_power')) == args.argument_main) 
            return Promise.resolve(result);
        });
        this.pollSeverCloud(settings);
	}

    conditionIsGeneratingToString(power) {
        if ( power < 100 ) {
            return 'minimal';
        } else if ( power < 500 ) {
            return 'ok';
        } else if ( power < 1000 ) {
            return 'better';
        } else if ( power < 1500 ) {
            return 'nice';
        } else if ( power < 2000 ) {
            return 'super';
        } else if ( power < 3000 ) {
            return 'chill';
        } else if ( power < 4000 ) {
            return 'awesome';
        } else {
            return 'moneymaker';
        }
    }

    // flow triggers
    triggerPowerAbove100WFlow(tokens) {
        this._flowTriggerPowerAbove100W
            .trigger(tokens)
            .then(this.log("triggerPowerAbove100WFlow"))
            .catch(this.error)
    }

    // flow triggers
    triggerPowerAbove500WFlow(tokens) {
        this._flowTriggerPowerAbove500W
            .trigger(tokens)
            .then(this.log("triggerPowerAbove500WFlow"))
            .catch(this.error)
    }
    // flow triggers
    triggerPowerAbove1000WFlow(tokens) {
        this._flowTriggerPowerAbove1000W
            .trigger(tokens)
            .then(this.log("triggerPowerAbove1000WFlow"))
            .catch(this.error)
    }

    triggerPowerIs0WFlow(tokens) {
        this._flowTriggerPowerIs0W
            .trigger(tokens)
            .then(this.log("triggerPowerIs0WFlow"))
            .catch(this.error)
    }

    onAdded() {
        let id = this.getData().id;
        this.log('device added: ', id);
    } // end onAdded

    onDeleted() {

        let id = this.getData().id;
        let name = this.getName() + '_' + this.getData().id;
        let cronName = name.toLowerCase();
        this.log('Unregistering cron:', cronName);
        Homey.ManagerCron.unregisterTask(cronName, function (err, success) {});

        this.log('device deleted:', id);

    } // end onDeleted

    onSettings(settings, newSettingsObj, changedKeysArr, callback) {
        try {
            for (var i = 0; i < changedKeysArr.length; i++) {
                switch (changedKeysArr[i]) {
                    case 'apikey':
                        this.log('APIKey changed to ' + newSettingsObj.apikey);
                        settings.apikey = newSettingsObj.apikey;
                        break;
                    case 'appkey':
                        this.log('APPKey changed to ' + newSettingsObj.appkey);
                        settings.appkey = newSettingsObj.appkey;
                        break;
                    case 'secret':
                        this.log('Secret changed to ' + newSettingsObj.secret);
                        settings.secret = newSettingsObj.secret;
                        break;
                    default:
                        this.log("Key not matched: " + i);
                        break;
                }
            }

            let name = 'zeversolar_plant_' + this.getData().id;
            let cronName = name.toLowerCase();
            this.log("device settings " +  JSON.stringify(settings));
    
            Homey.ManagerCron.getTask(cronName)
                .then(task => {
                    this.log("The task exists: " + cronName);
                    this.log('Unregistering cron:', cronName);
                    Homey.ManagerCron.unregisterTask(cronName, function (err, success) {});
                    Homey.ManagerCron.registerTask(cronName, "*/5 * * * *", settings)
                    .then(task => {
                        task.on('run', settings => this.pollSeverCloud(settings));
                    })
                    .catch(err => {
                        this.log('problem with registering cronjob: ${err.message}');
                    });            
                })
                .catch(err => {
                    if (err.code == 404) {
                        this.log("The task has not been registered yet, registering task: " + cronName);
                        Homey.ManagerCron.registerTask(cronName, "*/5 * * * *", settings)
                            .then(task => {
                                task.on('run', settings => this.pollSeverCloud(settings));
                            })
                            .catch(err => {
                                this.log('problem with registering cronjob: ${err.message}');
                            });
                    } else {
                        this.log('other cron error: ${err.message}');
                    }
                });

            this.pollSeverCloud(settings);
            callback(null, true)
        } catch (error) {
            callback(error, null)
        }
    }


    pollSeverCloud(settings) {
        this.log("pollSeverCloud settings " +  JSON.stringify(settings));
        // let settings2 = this.getSettings();
        let name2 = 'zeversolar_plant_' + this.getData().id;
        this.log("pollSeverCloud device name " +  name2);
        // this.log("pollSeverCloud device settings " +  JSON.stringify(settings2));

        zevercloud.getTodayData(settings).then(data => {
            var currentdate =new Date().timeNow();
            this.log("refresh now " + currentdate);
    
            this.log("Received data");

            if (data != "ERROR"){

                this.log("object "+ JSON.stringify(data));
                var ludtdate = data.ludt; 
                this.log("update date "+ ludtdate);
                this.setCapabilityValue('latest_upload_date', ludtdate);

                var power = 0
                if ( data.Power.unit == "kW" ) {
                    power =  (data.Power.value * 1000);
                } else {
                    power =  data.Power.value;
                }

                var etotal = 0
                if ( data["E-Total"].unit == "MWh" ) {
                    etotal =  (data["E-Total"].value * 1000);
                } else {
                    etotal =  data["E-Total"].value;
                }

                var etotalMonth = 0
                if ( data["E-Month"].unit == "MWh" ) {
                    etotalMonth =  (data["E-Month"].value * 1000);
                } else {
                    etotalMonth =  data["E-Month"].value;
                }

                this.setCapabilityValue('measure_power', power);

                var val = data["E-Today"].value
                this.setCapabilityValue('measure_e-total-today', val);
                this.setCapabilityValue('measure_e-total-month', etotalMonth);
                this.setCapabilityValue('measure_e-total', etotal);

                let tokens = {
                    "power": power,
                    "plant": this.getData().id.toLowerCase()
                };

                if ( this.getCapabilityValue('measure_power') < 100 
                    && power > 100 
                    && power < 500 ) {
                    this.triggerPowerAbove100WFlow(tokens);
                } else if ( this.getCapabilityValue('measure_power') < 500 
                            && power > 500 
                            && power < 1000 ) {
                    this.triggerPowerAbove500WFlow(tokens);
                } else if ( this.getCapabilityValue('measure_power') < 1000 
                            && power > 1000 ) {
                    this.triggerPowerAbove1000WFlow(tokens);
                }

                if (this.getCapabilityValue('measure_power') >= 25 && power < 25 ) {
                    let tokens2 = {
                        "power": 0,
                        "plant": this.getData().id.toLowerCase()
                    };
                    this.triggerPowerIs0WFlow(tokens2);
                }
            } else {
                this.setCapabilityValue('latest_upload_date', "Err check your keys");
            }
        })
        .catch(error => {
            this.log(error);
        });
    }

}

module.exports = plant;